pkg/asset/agent/image/agentimage.go

```diff
@@ -103,29 +133,73 @@ func (a *AgentImage) updateIgnitionImg(ignition []byte) error {
 		return err
 	}
 
-	ignitionImgPath := filepath.Join(a.tmpPath, "images", "ignition.img")
-	fi, err := os.Stat(ignitionImgPath)
-	if err != nil {
-		return err
+	// Gathering the ignition image details
+	var (
+		ignImagePath                 string
+		ignStartOffset, ignMaxLength int64
+	)
+	ignInfoJSONPath := filepath.Join(a.tmpPath, "coreos", "igninfo.json")
+	if _, err := os.Stat(ignInfoJSONPath); err == nil {
+		// Fetching the ignition details from the igninfo file
+		err = a.fetchIgnitionInfo(ignInfoJSONPath)
+		if err != nil {
+			return err
+		}
+		ignImagePath = filepath.Join(a.tmpPath, ignInfo.File)
+		ignStartOffset = ignInfo.Offset
+		ignMaxLength = ignInfo.Length
+
+	} else {
+		// Defaulting the ignition image path if no info file is found
+		ignImagePath = filepath.Join(a.tmpPath, "images", "ignition.img")
+		ignStartOffset = 0
+		ignMaxLength = 0
 	}
 
-	// Verify that the current compressed ignition archive does not exceed the
-	// embed area (usually 256 Kb)
-	if len(ignitionBuff) > int(fi.Size()) {
-		return fmt.Errorf("ignition content length (%d) exceeds embed area size (%d)", len(ignitionBuff), fi.Size())
+	// Verify that the compressed ignition buffer archive does not exceed the embed area (usually 256 Kb)
+	if ignMaxLength == 0 {
+		// Checking the file size if no ignition length is provided
+		fileInfo, err := os.Stat(ignImagePath)
+		if err != nil {
+			return err
+		}
+
+		if len(ignitionBuff) > int(fileInfo.Size()) {
+			return fmt.Errorf("ignition content length (%d) exceeds embed area size (%d)", len(ignitionBuff), fileInfo.Size())
+		}
+	} else if int64(len(ignitionBuff)) > ignMaxLength {
+		return fmt.Errorf("Ignition content length (%d) exceeds embed area size (%d)", len(ignitionBuff), ignMaxLength)
 	}
 
-	ignitionImg, err := os.OpenFile(ignitionImgPath, os.O_WRONLY, 0)
+	// Open the image file with permissions to seek and write the ignition content
+	ignitionImg, err := os.OpenFile(ignImagePath, os.O_RDWR, 0644)
 	if err != nil {
 		return err
 	}
 	defer ignitionImg.Close()
 
+	// Adjusting the ignition image file to the offset of ignition content (in case of cdboot.img)
+	_, err = ignitionImg.Seek(ignStartOffset, 0)
+	if err != nil {
+		return err
+	}
+
+	// Write the ignition content into the image file
 	_, err = ignitionImg.Write(ignitionBuff)
 	if err != nil {
 		return err
 	}
 
+	// Padding 0's at the end if the ignition content has less bytes in the image file (in case of cdboot.img)
+	paddingLength := ignMaxLength - int64(len(ignitionBuff))
```

Comment:
I don't know that this is actually required; there is nothing here that truncates the file as far as I can see.

Reply:
@zaneb , We are doing this not for the image file but to maintain the integrity of the ignition byte content just to avoid having any garbage values getting into it if the length of the ignition buffer is lesser than the maximum length it should be

Reply:
Not sure if for s390x case is different, but isn't the original archive already fully 0-padded? Overwriting the current ignition buffer wouldn't be sufficient to preserve the (already existing) padding?

Reply:
@phani2898 better to remove the padding logic.

Reply:
removed padding logic as suggested.